common.py

```diff
@@ -90,41 +92,101 @@ def str_to_list(input_str: str) -> list[int]:
 
     for part in parts:
         if '-' in part:
-            start, end = map(int, part.split('-'))
-            result.update(range(start, end + 1))
+            start, end = part.split('-')
+            result.update(range(int(start), int(end) + 1))
         else:
             result.add(int(part))
 
     return sorted(result)
 
 
+@dataclass(frozen=True, init=False)
 class RangeList:
-    _range: list[tuple[bool, list[int]]] = []
-    initial_values: Optional[list[int]] = None
+    _include: Optional[set[int]]
+    _exclude: Optional[set[int]]
 
-    def __init__(self, initial_values: Optional[list[int]] = None):
-        self.initial_values = initial_values
+    UNLIMITED: typing.ClassVar['RangeList']
 
-    def _append(self, lst: list[int], expand: bool) -> None:
-        self._range.append((expand, lst))
+    ACCUMULATOR_START: tuple[Optional[set[int]], Optional[set[int]]] = (None, None)
 
-    def include(self, lst: list[int]) -> None:
-        self._append(lst, True)
+    def __init__(
+        self,
+        include: Optional[Iterable[int]] = None,
+        exclude: Optional[Iterable[int]] = None,
+    ):
+        def prepare(arg: Optional[Iterable[int]]) -> Optional[set[int]]:
+            return set(arg) if arg is not None else None
 
-    def exclude(self, lst: list[int]) -> None:
-        self._append(lst, False)
+        object.__setattr__(self, "_include", prepare(include))
+        object.__setattr__(self, "_exclude", prepare(exclude))
 
-    def filter_list(self, initial: list[T]) -> list[T]:
-        applied = set(range(len(initial)))
-        if self.initial_values is not None:
-            applied &= set(self.initial_values)
+    def match(self, idx: int) -> bool:
+        if self._include is not None and idx not in self._include:
+            return False
+        if self._exclude is not None and idx in self._exclude:
+            return False
+        return True
+
+    def filter(self, lst: Iterable[T]) -> list[T]:
+        lst = list(lst)
+        return [lst[idx] for idx in range(len(lst)) if self.match(idx)]
+
+    @staticmethod
+    def parse_accumulate(
+        range_accumulator: tuple[Optional[set[int]], Optional[set[int]]],
```

Comment:
This looks quite complex. Can we simplify?

Reply:
changed. Better?

The `RangeList` is now the accumulator state, and no longer has it in a separate tuple. That makes the `RangeList._accumulate()` method mutate the object, but that seems fine here.